PR: https://github.com/odoo/odoo/pull/

From: cd4b10b004f69098b6cf8193a171bb3c14f22b97
From: Joren Van Onder
Date: 2019-08-23 20:33:28

Structural Changes: 4
Total Changes: 12

[IMP] point_of_sale: generic electronic payment interface

This is an interface that can be implemented to support some
electronic payment method in the POS. It's extracted from the existing
pos_iot code.

To implement a payment method you:

- Python: add a selection to the use_payment_terminal field in Python
- JS: implement the PaymentInterface class
- JS: register your PaymentInterface in point_of_sale.models:

models.register_payment_method(
  'use_payment_terminal_selection_value', ImplementedPaymentInterface
);

See the documentation of PaymentInterface for details on what each
method does and when it is called.

Co-authored-by: Antoine Prieels <anp@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/pos_order.py
+++ b/addons/point_of_sale/models/pos_order.py
@@ -59,6 +59,8 @@ class PosOrder(models.Model):
             'amount': ui_paymentline['amount'] or 0.0,
             'payment_date': payment_date,
             'payment_method_id': ui_paymentline['payment_method_id'],
+            'card_type': ui_paymentline.get('card_type'),
+            'transaction_id': ui_paymentline.get('transaction_id'),
             'pos_order_id': order.id,
         }
 

--- a/addons/point_of_sale/models/pos_payment.py
+++ b/addons/point_of_sale/models/pos_payment.py
@@ -23,3 +23,5 @@ class PosPayment(models.Model):
     partner_id = fields.Many2one('res.partner', string='Customer', related='pos_order_id.partner_id')
     session_id = fields.Many2one('pos.session', string='Session', related='pos_order_id.session_id')
     company_id = fields.Many2one('res.company', string='Company', related='pos_order_id.company_id')
+    card_type = fields.Char('Type of card used')
+    transaction_id = fields.Char('Payment Transaction ID')

--- a/addons/point_of_sale/models/pos_payment_method.py
+++ b/addons/point_of_sale/models/pos_payment_method.py
@@ -40,6 +40,14 @@ class PosPaymentMethod(models.Model):
     open_session_ids = fields.Many2many('pos.session', string='Pos Sessions', compute='_compute_open_session_ids', help='Open PoS sessions that are using this payment method.')
     config_ids = fields.Many2many('pos.config', string='Point of Sale Configurations')
     company_id = fields.Many2one('res.company', string='Company', default=lambda self: self.env.company)
+    use_payment_terminal = fields.Selection([], 'Use a Payment Terminal', help='Record payments with a terminal on this journal.')
+    hide_use_payment_terminal = fields.Boolean(compute='_compute_hide_use_payment_terminal', help='Technical field which is used to '
+                                               'hide use_payment_terminal when no payment interfaces are installed.')
+
+    def _compute_hide_use_payment_terminal(self):
+        no_terminals = not bool(self._fields['use_payment_terminal'].selection)
+        for payment_method in self:
+            payment_method.hide_use_payment_terminal = no_terminals
 
     @api.depends('config_ids')
     def _compute_open_session_ids(self):
